src/transform/src/predicate_pushdown.rs

```diff
@@ -77,8 +77,29 @@ impl crate::Transform for PredicatePushdown {
 }
 
 impl PredicatePushdown {
-    /// Pushes predicates down through other operators.
-    pub fn action(
+    /// Pushes predicates down through the operator tree and extracts
+    /// The ones that should be pushed down to the next dataflow object
+    pub fn dataflow_transform(
+        &self,
+        relation: &mut RelationExpr,
+        get_predicates: &mut HashMap<Id, HashSet<ScalarExpr>>,
+    ) {
```

Comment:
I'm not certain I understand what is happening here! :D

I understand that calling `action` previously wouldn't have been recursive, but I think the fix (maybe wrong) it to call a modified `transform` instead, rather than this new method. In particular, since `dataflow_transform` doesn't seem to call `action`, if predicates are pushed in to a dataflow from outside, they won't propagate through the dataflow to its inputs.

I recommend (though if you disagree, push back) a method that is just `transform` with the `HashMap` exposed, so that it can be pre-populated and have its contents read out at the end.

Reply:
I previously had `dataflow_transform` recursively call `action`, but it resulted in issues because of a filter being pushed into a join whose implementation had already been planned. 

An example:
```
SELECT x29,x31,x51,x55
  FROM t31,t55,t29,t51
 WHERE b55=a31
   AND a51=b31
   AND a29=6
   AND b51=a29
```

`EXPLAIN PLAN FOR` the query results in this:

```
%0 =
| Get materialize.public.t31 (u10)
| ArrangeBy (#0)

%1 =
| Get materialize.public.t55 (u14)

%2 =
| Get materialize.public.t29 (u8)
| ArrangeBy (#0)

%3 =
| Get materialize.public.t51 (u12)
| ArrangeBy (#0)

%4 =
| Join %0 %1 %2 %3 (= #0 #4) (= #1 #9) (= #6 #10)
| | implementation = Differential %1 %0.(#0) %3.(#0) %2.(#0)
| | demand = (#2, #5, #6, #8, #11)
| Filter (#6 = 6)
| Project (#8, #2, #11, #5)
```

Having `dataflow_transform` recursively call `action` results in a plan that looks like this. The filter has been pushed into the join as an equivalence class. The filter does not make it further down because of the ArrangeBy's. 

```
%0 =
| Get ? (u10)
| ArrangeBy (#0)

%1 =
| Get ? (u14)

%2 =
| Get ? (u8)
| ArrangeBy (#0)

%3 =
| Get ? (u12)
| ArrangeBy (#0)

%4 =
| Join %0 %1 %2 %3 (= #0 #4) (= #1 #9) (= #6 #10) (= #6 6)
| | implementation = Differential %1 %0.(#0) %3.(#0) %2.(#0)
| | demand = (#2, #5, #6, #8, #11)
| Filter !(isnull(#6)), !(isnull(6))
| Project (#8, #2, #11, #5)
```

There seems to be an issue in the join rendering that causes the `(=#6 6)` equivalence class to not run, so this change causes some sqlite tests to return incorrect results.

I can look into the join rendering issue, but what do you think we should do about the filter suboptimally being pushed down into the join?

Reply:
That's a great point!

I suspect that we'll eventually need to run almost all of the optimizations again. At least, some folks will happily use views to create joins and then hit them with constraints in follow-on views. We could announce that as an anti-pattern, but probably the right thing is to push all optimizations down further.

Let's see, for the moment let's not do that because that's a lot for this PR. But we should make a note about fixing it!